types/react/index.d.ts

```diff
@@ -496,12 +486,7 @@ declare namespace React {
         forceUpdate(callback?: () => void): void;
         render(): ReactNode;
 
-        // React.Props<T> is now deprecated, which means that the `children`
-        // property is not available on `P` by default, even though you can
-        // always pass children as variadic arguments to `createElement`.
-        // In the future, if we can define its call signature conditionally
-        // on the existence of `children` in `P`, then we should remove this.
-        readonly props: Readonly<P> & Readonly<{ children?: ReactNode | undefined }>;
+        readonly props: Readonly<P>;
```

Comment:
Probably need to undo this moment, wether a user or lib specify this, changes nothing. All it's doing is forcing everyone in the eco who rely on this (and proven by the countless of prs already raised).

In the future when execution pattern changes, would make a change like this more favourable, as there is underlying breaking changes as well.

- [ ] undo me

Reply:
> The children are sent by react, why must the user define this?!

The component implementation takes care of this.

It sounds like you would also argue that any excess prop should be allowed since React "sends it".

Take this component implementation:

```tsx
import * as React from 'react';

interface ComponentProps {

}

const Component: React.FunctionComponent<ComponentProps> = () => {
  return <input />
}

// we allow this
<Component>foo</Component>;
// but not this
<Component children2="foo" />;
// but also this
<Component children="foo" />;

// Why do we only reject `children2` when `Component` does neither implement `children` nor `children2`
```

[Playground Link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdwAdjFlAXlnAMImRNYWABWJh0AbwYBfBrghNU8XuHmCYALkxUYAOgBiAVyb5g85fzUAec6uGjUAPjgBeOAAoAlC6eS4cSjAGUExwlsxgBvAA9A50MnRRUXAA7lzIADbpEMlwMAAWwKh01ny2MA4EEBCWUTYCLA70iXAARpFwTBDw+YXFdWpwuAXpACaUTABMzgBElRDTcDFNSW3wGagQuQVFJSr18EPAo+MzcwtLDM0A6nkAnnAjm6lw8un3lABWWPhwAAaHx0EE1+KTygj+-RYIMeWHQAmA+TYcFAYHSWBAA3+wzGghBnSgfwBOMmvyAA)

> In the future when execution pattern changes, would make a change like this more favourable, as there is underlying breaking changes as well.

I don't understand what this means. Could you rephrase this?

> All it's doing is forcing everyone in the eco who rely on this (and proven by the countless of prs already raised) to define this in user land 🫠 which defeats the point of separation of concerns. 

While also catching a lot of existing buts and fixing use cases where implicit children are undesired (e.g. when they use render props or more specific children with `ReactElement`).
It's not a blanket "we move this to userland". That would be silly. This is an actual bug in the types that we fix that many existing types rely on. This is why this is also a breaking change. In an ideal world nobody would've relied on it, but that's just not realistic with any implicit behavior.

Reply:
All components have children, until that changes upstream there is no reason to remove it now. You're hurting the eco.

Reply:
> All components have children

Could you clarify what that means to you? 
Maybe start with this simple example here: `const Component = () => <input />`. Does `Component` "have children"?

>  until that changes upstream there is no reason to remove it now

What's "upstream" here?

Reply:
> Could you clarify what that means to you?

Yeah for sure;

```js
const A = () => <h1 />;
<A /> // has children

const B = ({children}) => <h1>{children}</h1>; // has children

<B>hi</B> // has children

...
```

whether you `createElement` with or without children, there is still a `children` prop passed.

[See here](https://codesandbox.io/s/wizardly-maxwell-o6iqpb)

notice how it has a child prob whether _I_ told it about it or not. Coz the types that currently exist in the dts, is right. Its [`ReactNode | null | undefined`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/03d5034a61b77fd7bb3fae09f26e1ffbbcedc32a/types/react/index.d.ts#L829)

your component will receive children in anycase. So i reiterate, until this changes on the upstream where _some_ components revive children, and others don't. These types as they stand are right.

---

> Maybe start with this simple example here: const Component = () => <input />. Does Component "have children"?

yes.

```jsx
const Component = () => <input />

<Component>hi eps1lon</Component>
```

both react runtime, and Typescript is happy here. React will give `Component` children, whether ts says it can or cant.

sometimes you wanna trick the system, and let ts know, please dont let my consumers send children (even though its still there).

```js
// ❌ typescript will hate you, but will run 100% fine.
const Component:NewFunctionComponentWithoutChildren = ({children}) => <h1>{children}</h1>

<Component>hi 1ps1lon</Component>
```

and because it will run 100% fine, and is what the native behavior is of the upstream, the types should reflect that. For one to deviate from whats native, aka your `input` example.

opt out like -> `const Component:FC<{children?: never}> = () => <input />`

you now have a far better api, one that will disguise what react is doing under the hood, but the consumer chose to deviate there.

---

> What's "upstream" here?

github.com/facebook/react

Reply:
> [See here](https://codesandbox.io/s/wizardly-maxwell-o6iqpb)

>
> notice how it has a child prop whether I told it about it or not

I'm not seing this in your codesandbox. I get the following:

```
{}

{"children":"hi"}

{}

{"children":"hi"}
```
There's no `children` when you didn't pass it.

> > What's "upstream" here?
>
> github.com/facebook/react

This change is not targetting host components that are implemented by `react-dom` or other renderes. This change is targetting user-land components. They decide whether they handle children. Not React.

> both react runtime, and Typescript is happy here

1. TypeScript is happy here because we have implicit children. This is not an argument here. Otherwise any change is defeated by that argument because it changes the status quo.
2. React would be happy about any excess prop. Yet TypeScript rejects any other excess prop but `children`. You have yet to show why rejecting excess props is fine for **every** prop but `children`.

Reply:
Probably we have to agree to disagree here mate. Children is special as it comes from react, in the 99% case. And for a user to tell react a prop it itself controls, is a separation of concern we shouldn't cross in user land. 

Let's leave it here, and see what other folk have to say.

Reply:
Regarding deferring to facebook/react here:
The flow types in facebook/react do not add implicit children:
- [React.Component](https://github.com/facebook/flow/blob/ec71532fb35859bc46644ef9903764067c93f44a/lib/react.js#L43)
- https://github.com/facebook/flow/blob/ec71532fb35859bc46644ef9903764067c93f44a/lib/react.js#L43

So upstream is an argument for removing implicit children.